app/controllers/users/registrations_controller.rb

```diff
@@ -1,11 +1,28 @@
-# frozen_string_literal: true
-
 class Users::RegistrationsController < Devise::RegistrationsController
 
+  def create
+    if email_matches_whitelist
+      super
+    else
+      build_resource
```

Comment:
This also feels like a devise-specific thing that might not make sense to a reader. More inclined to pull this out and explain it through method naming.